Implement `bal test --cloud` for executing tests in docker containers

Ramith-D-Rodrigo opened the pull request:
## Purpose
To execute ballerina project tests in docker container in order to verify the business logic functionality within these said containers. To this end, a new flag named "--cloud" for `bal test` command is required.

Fixes #41036

## Approach
Creates a standalone fat jar that contains all the test modules and the relevant test execution dependencies. 
Once created, the C2C lifecycle plugin will notify the C2C for generating the docker artifacts. Since the test execution is done at C2C side, the RunTestsTask will not be executed. Hence, features such as code coverage and test reports are not supported.

Currently, implemented only for the JVM path.

Unit tests will be made after the initial code review.

## Samples
> Provide high-level details about the samples related to this feature.

## Remarks
~~TODO: implement the GraalVM path execution.~~

After 4084925ebf344942adb0d89da93401abb4bc5ffc commit (including the previous commits that were done recently), GraalVM path execution works for the cloud test execution.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

ShammiL commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/CustomSystemClassLoader.java:
```diff
@@ -0,0 +1,116 @@
+
+import io.ballerina.projects.util.ProjectConstants;
+
+import java.io.*;
```
It is a best practice to have each import separately than having them all together as `java.io.*`

Ramith-D-Rodrigo replied:
Done in [40229a4](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/40229a40c6e8385fc2d56f004746b7a5494411f1) commit.

ShammiL commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/CustomSystemClassLoader.java:
```diff
@@ -0,0 +1,116 @@
+        catch (NullPointerException | IOException e) {
+            throw new RuntimeException("Error reading excludingClasses.txt", e);
+        }
+
```
```suggestion
```

Ramith-D-Rodrigo replied:
Done in [40229a4](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/40229a40c6e8385fc2d56f004746b7a5494411f1) commit.

gayaldassanayake commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -326,6 +336,11 @@ public void execute() {
+        if (!project.buildOptions().cloud().isEmpty() && project.buildOptions().codeCoverage()) {
+            this.outStream.println("WARNING: Code coverage generation is not supported with Ballerina cloud test");
+        }
+
+
```
```suggestion
```

Ramith-D-Rodrigo replied:
Done in [40229a4](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/40229a40c6e8385fc2d56f004746b7a5494411f1) commit.

ShammiL commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/CustomSystemClassLoader.java:
```diff
@@ -0,0 +1,116 @@
+
+        return buff;
+    }
+}
```
There are redundant new lines used in most classes. You can check and remove them

ShammiL commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -336,19 +351,39 @@ public void execute() {
+                && createTestExecutableTask != null) {
+            taskBuilder.addTask(createTestExecutableTask);
+        }
+        //                .addTask(new CopyResourcesTask(), listGroups) // merged with CreateJarTask
```
Do we need this comment?

Ramith-D-Rodrigo replied:
This was there before the modification. While the comment does not match with the current implementation (the code structure), this task might have some relevance?

ShammiL commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:
```diff
@@ -163,7 +146,34 @@ public void execute(Project project) {
 
-    private void notifyPlugins(Project project, Target target) {
+    private Path getExecutablePath(Project project, Target target) {
+        Path executablePath;
+        try {
+            executablePath = target.getExecutablePath(project.currentPackage()).toAbsolutePath().normalize();
+        } catch (IOException e) {
+            throw createLauncherException(e.getMessage());
+        }
+        return executablePath;
```
```suggestion
        try {
            return target.getExecutablePath(project.currentPackage()).toAbsolutePath().normalize();
        } catch (IOException e) {
            throw createLauncherException(e.getMessage());
        }
```

Ramith-D-Rodrigo replied:
Done in [40229a4](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/40229a40c6e8385fc2d56f004746b7a5494411f1) commit.

ShammiL commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateTestExecutableTask.java:
```diff
@@ -0,0 +1,232 @@
+                    ZipFile zipFile = new ZipFile(moduleJarPath.toFile());
+
+                    zipFile.stream().forEach(entry -> {
+                        if (entry.getName().endsWith(".class")) {
+                            excludingClassPaths.add(entry.getName().replace("/", ".")
+                                    .replace(".class", ""));
```
Can use constants here. Check other places as well.

Ramith-D-Rodrigo replied:
Done in [40229a4](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/40229a40c6e8385fc2d56f004746b7a5494411f1) commit.

ShammiL commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateTestExecutableTask.java:
```diff
@@ -0,0 +1,232 @@
+                }
+
+                //create the single fat jar for all the test modules that includes the test suite json
+                EmitResult result = jBallerinaBackend.emit(
```
After a recent fix it was decided that the `EmitResult` diagnostics would only contain diagnostics reported in the executable emitting phase while `JBallerinaBackend` diagnostics would contain diagnostics relevant to all phases of the build. 
You can check https://github.com/ballerina-platform/ballerina-lang/pull/42131 to find these changes and make changes here appropriately.

Ramith-D-Rodrigo replied:
made the changes in [02decdf](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/02decdf78e7a2e1f4ab971a686a05d40d6b3b359) commit.

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -326,6 +336,11 @@ public void execute() {
         }
 
+        if (!project.buildOptions().cloud().isEmpty() && project.buildOptions().codeCoverage()) {
+            this.outStream.println("WARNING: Code coverage generation is not supported with Ballerina cloud test");
```
Let's test how `--cloud=k8s` is handled too. 
We have to issue a warning saying tests are only supported with `--cloud=docker`

Ramith-D-Rodrigo replied:
Will work on that. After issuing the warning, should we abort the execution?

Thevakumar-Luheerathan replied:
Either we need abort or skip the cloud flag and run normal tests IMO

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -336,19 +351,39 @@ public void execute() {
+        taskBuilder.addTask(cleanTargetCacheDirTask, isSingleFile) // clean the target cache dir(projects only)
+                .addTask(resolveMavenDependenciesTask) // resolve maven dependencies in Ballerina.toml
                 // compile the modules
-                .addTask(new CompileTask(outStream, errStream, false, isPackageModified, buildOptions.enableCache()))
```
shall we only add the Executable task addition without changing the existing implementation?

Ramith-D-Rodrigo replied:
By the existing implementation do you mean how these tasks are added to the TaskBuilder? Like how the Constructors are called within the addTask method? 

The only issue is that, the order matters with this approach and currently CreateTestExecutableTask requires the reference to the RunTestsTask to invoke required methods. Since, RunTestsTask happens after the creation of the test executable, we cannot construct this object after the CreateTestExecutableTask using this approach. 

However, currently RunTestsTask's execute method returns right after invocation if the cloud flag is enabled since we have delegated the execution of the tests to the C2C side. Therefore we do not actually need to add the RunTestsTask to the TaskBuilder. Hence, we can call RunTestsTask's constructor within CreateTestExecutableTask's constructor. Doing so, would cause readability issues. Furthermore, we also need to consider the cases where cloud flag is not given (Normal bal test execution).

Ramith-D-Rodrigo replied:
After the discussion, I have made the relevant changes to reflect the previous implementation while including the current one. The change can be found in [c329e24](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/c329e2498f275dc35210d6b116bdc585a851cf45) commit.

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateTestExecutableTask.java:
```diff
@@ -0,0 +1,232 @@
+import static io.ballerina.cli.launcher.LauncherUtils.createLauncherException;
+import static io.ballerina.projects.util.ProjectConstants.USER_DIR;
+
+public class CreateTestExecutableTask extends CreateExecutableTask {
```
What's the benefit of implementing the `CreateExecutableTask`?
IMO it's better to stick to the existing semantics of the tasks and implement `Task` interface instead.

Ramith-D-Rodrigo replied:
The benefits would be the similarities between these two classes which allowed reusing some methods and properties via inheritance. Should I move back to Task implementation? In that case, I would have to write some redundant code.

gayaldassanayake replied:
Does it make logical sense for the `CreateTestExecutableTask` to inherit the `CreateExecutableTask`? AFAIU these are two classes at the same level which do similar tasks. If you want to reuse the code, perhaps we can have an abstract class/ interface that is implemented by both these classes.

Ramith-D-Rodrigo replied:
Other than the re usability, there are no other reasons. 

> Does it make logical sense for the `CreateTestExecutableTask` to inherit the `CreateExecutableTask`? AFAIU these are two classes at the same level which do similar tasks. If you want to reuse the code, perhaps we can have an abstract class/ interface that is implemented by both these classes.

That seems to be the case, I only thought from the perspective of re usability. Upon checking the code, there's one function which I moved to an utility class and other functions which I redefined in the class again. I changed the class implementation to match the structure. These new changes can be found in 73fb681d399dd985d8d8bb8560fa216543b6b028 commit. Still believe there can be more improvements such as an interface like you've mentioned.

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java:
```diff
@@ -175,19 +222,91 @@ public void execute(Project project) {
+        }
+    }
+
+    public static boolean createTestSuiteIfHasTests(Project project, Target target,
```
```suggestion
    public static boolean updateTestSuitesForProject(Project project, Target target,
```

Ramith-D-Rodrigo replied:
Changed in 2b05eea commit.

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java:
```diff
@@ -175,19 +222,91 @@ public void execute(Project project) {
+        }
+    }
+
+    private void performTasksAfterTestCompletion(Project project, Target target, Path testsCachePath,
```
```suggestion
    private void performPostTestsTasks(Project project, Target target, Path testsCachePath,
```

Ramith-D-Rodrigo replied:
Done in [4968e6e](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/4968e6e4e65089855ddb48e35628b6b2d812786e) commit

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/TestUtils.java:
```diff
@@ -278,4 +300,163 @@ public static void cleanTempCache(Project project, Path cachesRoot) {
+        return cmdArgs;
+    }
+
+    public static void addOtherNeededArgs(List<String> cmdArgs, String target, String jacocoAgentJarPath,
```
```suggestion
    public static void appendRequiredArgs(List<String> cmdArgs, String target, String jacocoAgentJarPath,
```

Ramith-D-Rodrigo replied:
Changed in 2b05eea commit.

Dilhasha commented on compiler/ballerina-lang/src/main/java/io/ballerina/projects/ArtifactType.java:
```diff
@@ -0,0 +1,23 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Update the license header

Ramith-D-Rodrigo replied:
Done in [3d6d26f](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/3d6d26f6be6a80d0c512a58bb48b110408ad15ea) commit

Dilhasha commented on compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java:
```diff
@@ -547,6 +652,20 @@ private Path emitExecutable(Path executableFilePath) {
         return executableFilePath;
     }
 
+    private Path emitTest(Path executableFilePath, HashSet<JarLibrary> jarDependencies,
```
```suggestion
    private Path emitTestExecutable(Path executableFilePath, HashSet<JarLibrary> jarDependencies,
```

Ramith-D-Rodrigo replied:
Changed in 2b05eea commit.

Dilhasha commented on compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java:
```diff
@@ -547,6 +652,20 @@ private Path emitExecutable(Path executableFilePath) {
+            assembleTestExecutableJar(executableFilePath, manifest, jarDependencies, testSuiteJsonPath, jsonCopyPath,
+                    excludingClassPaths, classPathTextCopyPath);
+        } catch (IOException e) {
+            throw new ProjectException("error while creating the executable jar file for package '" +
```
```suggestion
            throw new ProjectException("error while creating the test executable jar file for package '" +
```

Ramith-D-Rodrigo replied:
Fixed in 2b05eea commit

Dilhasha commented on misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/TestProcessor.java:
```diff
@@ -147,30 +147,49 @@ private Map<Document, SyntaxTree> getTestSyntaxTreeMap(Module module) {
+            addTestExecutionDependencies(module, jarResolver, testSuite);
+        }
+
+        // TODO: Remove redundancy in addUtilityFunctions
```
Let's mention this in the issue and keep track.

Ramith-D-Rodrigo replied:
This TODO was already there before the modification

Dilhasha commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestMain.java:
```diff
@@ -95,9 +99,18 @@ public static void main(String[] args) throws IOException {
+            }
 
-            try (BufferedReader br = Files.newBufferedReader(testSuiteCachePath, StandardCharsets.UTF_8)) {
+            try (br) {
```
LEt's use try with resources

Ramith-D-Rodrigo replied:
Done in [b26b6e5](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/b26b6e545aebed16500e2530f29f5ba5253e6502) commit

Dilhasha commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaConstants.java:
```diff
@@ -101,4 +101,18 @@ public class TesterinaConstants {
 
+    //class to identify the indices of the BTestMain args
+    public static class RunTimeArgs {
+        public static final int IS_FAT_JAR_EXECUTION = 0;
```
Where are these being used?

Ramith-D-Rodrigo replied:
This class is used in C2C side. Specifically addDockerTestCMDArgs function in line no 344 of [DockerGenerator.java](https://github.com/ballerina-platform/module-ballerina-c2c/blob/6a9e1b23efcb508a7624bf0914e7e87f402205f5/compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java)

SasinduDilshara commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -336,19 +351,39 @@ public void execute() {
+        CreateTestExecutableTask createTestExecutableTask = null;
+
+        if (!project.buildOptions().cloud().isEmpty()) {
+            //if cloud flag is set, create the test executable
```
```suggestion
            // if cloud flag is set, create the test executable
```

Ramith-D-Rodrigo replied:
Fixed in [40229a4](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/40229a40c6e8385fc2d56f004746b7a5494411f1) commit.

SasinduDilshara commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/TestUtils.java:
```diff
@@ -278,4 +300,163 @@ public static void cleanTempCache(Project project, Path cachesRoot) {
+                }
+            }
+        }
+        dependencies = dependencies.stream().distinct().collect(Collectors.toList());
+        return dependencies;
```
```suggestion
        return dependencies.stream().distinct().collect(Collectors.toList());
```

Ramith-D-Rodrigo commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateExecutableTask.java:
```diff
@@ -54,9 +54,9 @@
-    private final transient PrintStream out;
-    private Path output;
-    private Path currentDir;
+    protected final transient PrintStream out;
```
This is done to extend this class

Ramith-D-Rodrigo commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java:
```diff
@@ -198,7 +317,6 @@ public void execute(Project project) {
             }
 
-            //Set 'hasTests' flag if there are any tests available in the package
             if (!hasTests) {
```
Changed in https://github.com/ballerina-platform/ballerina-lang/commit/2b05eea61b3d81d44facdb9a958f4593099a0089 commit.

gayaldassanayake commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaConstants.java:
```diff
@@ -101,4 +101,18 @@ public class TesterinaConstants {
+        public static final int SINGLE_EXEC_TESTS = 8;
+        public static final int IS_RERUN_TEST_EXECUTION = 9;
+    }
+
```
```suggestion
```

Ramith-D-Rodrigo replied:
Done in 9fe35ce commit

Thevakumar-Luheerathan commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateTestExecutableTask.java:
```diff
@@ -0,0 +1,224 @@
+    @Override
+    public void execute(Project project) {
+        this.out.println();
+
```
Remove new line here

gayaldassanayake commented on misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaConstants.java:
```diff
@@ -101,4 +101,18 @@ public class TesterinaConstants {
     public static final int IDENTIFIER_START_INDEX = 1;
     public static final int IDENTIFIER_END_INDEX = 5;
 
+    //class to identify the indices of the BTestMain args
```
Since this is a public class, let's have a documentation comment similar to TesterinaConstants class.

```
/**
 * RuntimeArgs identifies the indices of the BTestMain arg
 *
 * @since 2201.9.0
 */
```

Ramith-D-Rodrigo replied:
Added in 9fe35ce commit

gayaldassanayake commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateTestExecutableTask.java:
```diff
@@ -0,0 +1,219 @@
+        List<String> updatedSingleExecTests;
+        List<String> mockClassNames = new ArrayList<>();
+
+        boolean status = RunTestsTask.createTestSuitesForProject(project, target, testProcessor, testSuiteMap,
```
What does this `status` mean? Shall we have a better name that reflects the value it hold?

Ramith-D-Rodrigo replied:
Status indicates whether the test suite has been created if there are any tests. The value is false if no tests found. And yes, I Agree. I have done the relevant changes in 89c50f2 commit.

Thevakumar-Luheerathan commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -196,6 +197,14 @@ public TestCommand() {
             "generation")
     private String graalVMBuildOptions;
 
+    @CommandLine.Option(names = {"--output", "-o"}, description = "Write the output to the given file. The provided " +
```
```suggestion
    @CommandLine.Option(names = {"--output", "-o"}, description = "write the output to the given file. The provided " +
```

Ramith-D-Rodrigo replied:
Removed this flag as it is not needed (`--output`).

Thevakumar-Luheerathan commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -272,9 +280,18 @@ public void execute() {
         }
 
+        if (project.buildOptions().cloud().equals("k8s")) {
+            String errMsg = "Cloud artifact generation does not support Kubernetes for tests";
```
```suggestion
            String errMsg = "cloud artifact generation does not support Kubernetes for tests";
```

Ramith-D-Rodrigo replied:
Done

Thevakumar-Luheerathan commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CreateTestExecutableTask.java:
```diff
@@ -0,0 +1,417 @@
+import static io.ballerina.projects.util.ProjectConstants.BLANG_COMPILED_JAR_EXT;
+import static io.ballerina.projects.util.ProjectConstants.USER_DIR;
+
+public class CreateTestExecutableTask implements Task {
```
Add class level comments

Ramith-D-Rodrigo replied:
Added

Ramith-D-Rodrigo commented:
Changes have been done after the code review. Addressed the suggestions and added new code segments.

- Rebased the master (Not the latest anymore).
- Added a new task called "`CleanTargetBinTestsDirTask`" to clean previous `bal test --cloud=docker` results (the fat jar). 
- Added relevant test cases in `TestCommandTest` by mocking the cloud package.
- After the rebase, there's a runtime error when the project has mocking. This doesn't occur when the `--graalvm` option is given as the executing fat jar already contains the modified classes. Fixed this by reverting the `orderTests()`'s implementation from query action to foreach loop.

Ramith-D-Rodrigo commented:
Fixed the build tests failure and the strand dump test failure. Strand dump test failure was due to the change in execute.bal (using foreach instead of query action).

Ramith-D-Rodrigo commented:
Relevant C2C repo changes PR : https://github.com/ballerina-platform/module-ballerina-c2c/pull/765

Ramith-D-Rodrigo commented:
Still testing the Full build pipeline

Dilhasha reviewed (approved):
LGTM
